deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,239 @@
+
+    def dumps(self) -> str:
+        """Convert the pipeline to a YAML string.
+
+        :return: YAML string representation of the pipeline
+        """
+        ...
+
+    def add_component(self, name: str, instance: Any) -> None:
+        """Add a component to the pipeline.
+
+        :param name: Name of the component
+        :param instance: Component instance to add
+        """
+        ...
+
+
+def _enable_publish_to_deepset() -> None:
+    """Add publish methods to the Haystack Pipeline and AsyncPipeline classes.
+
+    This function is called by deepset_sdk.init() to set up the SDK.
+    Users should not call this function directly.
+    """
+    try:
+        from haystack import AsyncPipeline as HaystackAsyncPipeline
+        from haystack import Pipeline as HaystackPipeline
+    except ImportError as err:
+        raise ImportError("Can't import Pipeline or AsyncPipeline, because haystack-ai is not installed.") from err
+
+    async def publish_to_deepset_async(self: PipelineProtocol, config: IndexConfig | PipelineConfig) -> None:
+        """Publish index or pipeline to deepset AI platform asynchronously.
+
+        An index is a special type of pipeline with the purpose to preprocess files, preparing them for
+        search and store them in a document store.
+
+        :param config: Configuration for publishing, either `IndexConfig` or `PipelineConfig`.
+            If publishing an index, the config argument is expected to be of type `IndexConfig`,
+            if publishing a pipeline, the config argument is expected to be of type `PipelineConfig`.
+        """
+        api_config = CommonConfig()  # Uses environment variables
+        async with DeepsetCloudAPI.factory(api_config) as api:
+            service = PipelineService(api)
+            await service.publish_async(self, config)
+
+    def publish_to_deepset(self: PipelineProtocol, config: IndexConfig | PipelineConfig) -> None:
+        """Publish index or pipeline to deepset AI platform synchronously.
+
+        An index is a special type of pipeline with the purpose to preprocess files, preparing them for
+        search and store them in a document store.
+
+        :param config: Configuration for publishing, use either `IndexConfig` or `PipelineConfig`.
+            If publishing an index, the config argument is expected to be of type `IndexConfig`,
+            if publishing a pipeline, the config argument is expected to be of type `PipelineConfig`.
+        """
+        # creates a sync wrapper around the async method since the APIs are async
+        try:
+            loop = asyncio.get_event_loop()
+            should_close = False
+        except RuntimeError:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            should_close = True
+
+        try:
+            return loop.run_until_complete(publish_to_deepset_async(self, config))
+        finally:
+            if should_close:
+                loop.close()
```

Comment:
you are doing this instead of asyncio.run to get the event loop e.g. from jupyter, right ? (the other calls just wrap it in asyncio.run)